### DISCUSSION TITLE
Taskchampion server/storage integration, general questions

### DISCUSSION THREAD

**Discussion by aktaboot** [GothenburgBitFactory/taskwarrior#3226] [General]
Hello,

I have some questions regarding how taskchampion will integrate with taskwarrior 

To give some context, I am trying to build a rust application that would synchronize taskwarrior tasks with [etebase](https://www.etebase.com/)  Calendar+Tasks ( called Etesync).

I was thinking about parsing the exported json or the tw datafiles directly. But I came across the SQLite database and taskchampion integration.

As far as I understand, taskchampion will be responsible for all data storage, so I should be looking to sync with the taskchampion instance instead.

- How should third party app integration work with taskchampion ? I suppose through the sync server ?
- Is this currently doable under the develop branch ? 

Any hints on already available information that I may have missed are greatly appreciated 

Thanks a lot, love the project

**Comment by djmitche** [comment]
That's great! Early in the development of Taskchampion we actually looked at using etebase as a backend. I think it was a little experimental at the time, so we decided against it.

If I understand what you're suggesting, it's something very different: an app that will coordinate between tasks in Etesync and tasks in Taskwarrior. That's a great idea!

The idea is that integrations with Taskchampion run their own local replica, and share changes to the tasks with `replica.sync()`. So your interface to Taskchampion would be entirely local, via Rust APIs, and not directly via the sync server (but indirectly through `replica.sync()`).

This is doable on the develop branch.

I'd encourage you to have a look at `taskchampion/docs/src` and to run `rustdoc` for the `taskchampion` crate. Quite a bit has changed since the last time that crate was published! I'd expect we'll publish again when Taskwarrior is released.

**Reply by aktaboot** [reply]
Thanks a lot for the insightful reply, you got it right on what I am trying to achieve! 

I will be posting updates at some point when things start working!

**Reply by djmitche** [reply]
Sounds good. Let me know if you have questions!

**Comment by aktaboot** [comment]
Haven't made much progress with taskwarrior integration as I figured I'd wait for it the next release and publishing of the taskchampion-lib crate, but my problem might be unrelated

I had some issues while importing the lib crate, I could not use Replica or Storage to access the sqlite database.

for example if I try to `use taskchampion_lib::replica::Replica` : Struct Replica is private 
I am also not sure how I should go about accessing the Storage

My initial thought was I would use `SqliteStorage::new(sqlite.db)` to access the database, not sure what I am doing wrong

**Reply by djmitche** [reply]
I think you'll want to be using `taskchampion` -- the `taskchampion-lib` crate just exposes the C API, which is not especially easy to use from Rust.

It sounds like you have the right idea regarding storage -- just the wrong crate :)
